### PR TITLE
KeywordBear.py: Output appropriate message

### PR DESCRIPTION
**Pull request by Techievena**
Output appropriate message if the language given in input is not valid/not supported for KeywordBear

Fixes https://github.com/coala/coala-bears/issues/1256

### CODE REVIEW

**Comment by Shade5** [comment]
KeywordBear.py is not fully covered by the tests. You'll have to add tests for 100% coverage

**Comment by Shade5** [comment]
Run `py.test --cov` to check coverage. You have get 100% coverage for KeywordBear.py

File: bears/general/KeywordBear.py [hunk 1]
```diff
@@ -18,8 +18,12 @@ def _get_comments(dependency_results):
             not isinstance(annotation_bear_results, list)):
         return
 
-    for result in annotation_bear_results:
-        yield from result.contents.get('comments', [])
+    try:
+        for result in annotation_bear_results:
+            yield from result.contents.get('comments', [])
+    except AttributeError:
```
Comment (aptrishu) [inline bears/general/KeywordBear.py]: check [January 7, 2017 7:17 PM](https://gitter.im/coala/coala?at=5870f1855ffdeea7231a007e)

Comment (sils) [inline bears/general/KeywordBear.py]: @aptrishu I think it's a lot simpler to use github as the main review UI, just copying the text from gitter makes it way easier for other reviewers to see what's going on here

File: bears/general/KeywordBear.py [hunk 2]
```diff
@@ -92,20 +97,21 @@ def run(self,
...
-        yield from self.check_keywords(filename, file, comments,
-                                       simple_keywords_regex, message)
-
-        if regex_keyword is not '':
-            regex = re.compile(regex_keyword)
-            message = ("The line contains the keyword '{}' which "
-                       'resulted in a match with given regex.')
-            yield from self.check_keywords(filename, file, comments, regex,
-                                           message)
+        if comments:
```
Comment (sils) [inline bears/general/KeywordBear.py]: I'd just do `if not comments: return` so you don't have to indent everything below

Comment (sils) [inline bears/general/KeywordBear.py]: well actually we still want the keywordbear to run even if it doesn't know where the comments are! Knowing the comments is optional and if they're not there the bear  should still work!

File: tests/general/KeywordBearTest.py [hunk 3]
```diff
@@ -214,3 +214,20 @@ def test_keyword_regex(self):
...
+        section.append(Setting('language', 'anything'))
+        section.append(Setting('keywords', 'TODO'))
+        uut = KeywordBear(section, Queue())
+        dep_results = {
+            'AnnotationBear': HiddenResult(
+                'AnnotationBear',
+                'coalang specification for anything not found.')
+        }
+
+        text = ['a==b']
```
Comment (sils) [inline tests/general/KeywordBearTest.py]: put a todo in here and make asserts that the bear still works even if the language is unknown, we just want the warning right?

Comment (Techievena) [inline tests/general/KeywordBearTest.py]: Thanks @sils for your valuable suggestions.  Actually I had wrong concepts about working of Keyword Bear. I was therefore completely pissed off. Well a question for now. Does Keyword Bear have to return the position of  keywords in comments only or it returns position of keywords in the code too?

**Comment by sils** [comment]
all keywords, also if its in the code

> I was therefore completely pissed off

did we do something wrong? If so please tell us so we know what we can do
better next time!

Sincerely,

Lasse Schuirmann

lasse@schuirmann.net
http://coala.io/ - http://viperdev.io/ - http://gitmate.com/

File: bears/general/KeywordBear.py [hunk 4]
```diff
@@ -92,6 +97,9 @@ def run(self,
         '''
         comments = _get_comments(dependency_results)
 
+        if not comments:
+            return
```
Comment (sils) [inline bears/general/KeywordBear.py]: sorry but this means that if the language isn't known the bear won't even look for keywords

Comment (Techievena) [inline bears/general/KeywordBear.py]: Well it is looking for them inspite of these statements.
image

Comment (Techievena) [inline bears/general/KeywordBear.py]: I am removing them as they are unnecessary.

File: bears/general/KeywordBear.py [hunk 5]
```diff
@@ -18,8 +19,12 @@ def _get_comments(dependency_results):
             not isinstance(annotation_bear_results, list)):
         return
 
-    for result in annotation_bear_results:
-        yield from result.contents.get('comments', [])
+    try:
```
Comment (sils) [inline bears/general/KeywordBear.py]: why is this needed? If the comments are unavailable that should be atched by the ifs above

Comment (sils) [inline bears/general/KeywordBear.py]: does this even have any effect?

**Comment by Techievena** [comment]
What am I missing?

**Comment by Shade5** [comment]
Did you figure out how to write the tests or do you need help with that?

**Comment by Techievena** [comment]
@Shade5 Need help in writing test case

**Comment by Shade5** [comment]
Does `py.test` work fine without your changes?
Show me the output of `py.test -k KeywordBearTest --cov`

**Comment by Techievena** [comment]
Here is it @Shade5 
image

**Comment by Shade5** [comment]
So you need to write two tests. One that satisties the `if` at 23 and another that does not. Add those tests to KeywordBearTest.py

**Comment by Techievena** [comment]
Yes I am trying to cover the log statements using assertLogs but they are failing. Can you have a look on test case I have wrriten

**Comment by Shade5** [comment]
How are you calling `_get_comments()` in your test?

**Comment by Techievena** [comment]
I think `execute_bear` is calling them.

**Comment by Shade5** [comment]
Any errors with `py.test -k KeywordBearTest --cov` ?

**Comment by Shade5** [comment]
I'm geting `logging not definied`

**Comment by Techievena** [comment]
I've imported logging in both Keyword.py and KeywordBearTest.py and `py.test -k KeywordBearTest --cov` is returning the same as before

**Comment by Techievena** [comment]
image

File: tests/general/KeywordBearTest.py [hunk 6]
```diff
@@ -214,3 +214,22 @@ def test_keyword_regex(self):
...
                                                 'in a match with given regex.')
+
+    def test_wrong_language(self):
+        self.section.append(Setting('language', 'anything'))
+        annotation_bear_result_type = namedtuple('result', 'contents')
+        dep_results = {
+            'AnnotationBear': [
+                annotation_bear_result_type(
+                  'coalang specification for anything not found.'
+                  )
```
Comment (adtac) [inline tests/general/KeywordBearTest.py]: We end the brackets/parantheses in the same line as the last element. See line 207 of this file.

File: tests/general/KeywordBearTest.py [hunk 7]
```diff
@@ -214,3 +214,22 @@ def test_keyword_regex(self):
...
+            ]
+        }
+
+        text = ['# todo 123']
+
+        with execute_bear(self.uut, 'F', text,
```
Comment (adtac) [inline tests/general/KeywordBearTest.py]: Please use named arguments - they're easier to read. Every other test uses them and consistent is better :)

See line 211 of this file.

File: tests/general/KeywordBearTest.py [hunk 8]
```diff
@@ -214,3 +214,22 @@ def test_keyword_regex(self):
...
+                          dependency_results=dep_results) as result:
+            self.assertEqual(result[0].diffs, {})
+            self.assertEqual(result[0].affected_code[0].start.line, 1)
+            self.assertEqual(len(result), 1)
```
Comment (adtac) [inline tests/general/KeywordBearTest.py]: this should probably be the first assert :3

Comment (adtac) [inline tests/general/KeywordBearTest.py]: And there aren't any checks to make sure the message is logged? I mean, that's what the whole PR is about.

**Comment by Techievena** [comment]
@adtac Done with the changes. Is it mergeable now? :smiley:

File: bears/general/KeywordBear.py [hunk 9]
```diff
@@ -19,7 +20,11 @@ def _get_comments(dependency_results):
         return
 
     for result in annotation_bear_results:
-        yield from result.contents.get('comments', [])
+        if isinstance(result.contents, str):
+            logging.error(result.contents)
+            yield []
```
Comment (sils) [inline bears/general/KeywordBear.py]: why do you yield an empty list? I'd just remove it.

Comment (adtac) [inline bears/general/KeywordBear.py]: what if I'm expecting a list? not yielding a list might trigger an exception

Comment (Techievena) [inline bears/general/KeywordBear.py]: As it was yeilding an empty list when `comments` key was not found in request.contents I thought I had to do the same in case when this execption arises.
Fine I'll remove it :+1:

Comment (sils) [inline bears/general/KeywordBear.py]: uhm yield is not a return @adtac

**Comment by sils** [comment]
ack https://github.com/coala/coala-bears/pull/1271/commits/75ae1c9d38ab26533838f852c7d9801e6e6af488

**Comment by sils** [comment]
@rultor merge

File: bears/general/KeywordBear.py (hunk 9 above)
Comment (adtac) [inline bears/general/KeywordBear.py]: ah didn't realize the original was a `yield from`, not a `yield`
